### DISCUSSION TITLE
Is the device applicable?

### DISCUSSION THREAD

**Discussion by qianzhang1994** [alan-turing-institute/PeriPy#117] [Q&A]
There is my PC device, is it unavailable for the program? The exclusion results show it does not support double precious.

print(cl.get_platforms())
print(cl.device_type.ALL)
print(cl.device_type.GPU)
print(platform.get_devices(device_type))
print(cl.Context([device]))

[<pyopencl.Platform 'Intel(R) OpenCL HD Graphics' at 0x19ac597c890>]
4294967295
<pyopencl.Context at 0x19ad4f930f0 on <pyopencl.Device 'Intel(R) Iris(R) Xe Graphics' on 'Intel(R) OpenCL HD Graphics' at 0x19ad52444d0>>
4
______________________________________
Device is Intel(R) Iris(R) Xe GraphicsGPU from Intel(R) Corporation with a max of 96 compute units,
a max of 256 work-items per work-group,
a max work item dimensions of 3,
a max work item sizes of [256, 256, 256],
and device local memory size is 65536 bytes.

**Comment by bb515** [comment]
Hello Quanzhang, thank you for your question.

What happens when you run example2 (the OpenCL example)? Could you post a full error message, or does the OpenCL code work just fine?

When you run the tests, via `pytest` in the repo directory, do the OpenCL integrator tests pass?

**Reply by qianzhang1994** [reply]
### There is a full error message.
Traceback (most recent call last):
  File "C:\Users\zhang\AppData\Local\Programs\Python\Python39\lib\site-packages\peripy-1.0.0-py3.9-win-amd64.egg\peripy\examples\example2\example
.py", line 252, in <module>
    main()
  File "C:\Users\zhang\AppData\Local\Programs\Python\Python39\lib\site-packages\peripy-1.0.0-py3.9-win-amd64.egg\peripy\examples\example2\example
.py", line 160, in main
    integrator = VelocityVerletCL(dt=dt, damping=damping)
  File "C:\Users\zhang\AppData\Local\Programs\Python\Python39\lib\site-packages\peripy-1.0.0-py3.9-win-amd64.egg\peripy\integrators.py", line 697
, in __init__
    super().__init__(*args, **kwargs)
  File "C:\Users\zhang\AppData\Local\Programs\Python\Python39\lib\site-packages\peripy-1.0.0-py3.9-win-amd64.egg\peripy\integrators.py", line 45,
 in __init__
    raise ContextError
peripy.integrators.ContextError: No suitable context was found. You can manually specifythe context by passing it to ModelCL with the 'context'ar
gument.

**Reply by qianzhang1994** [reply]
### There is the test information.
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master> peripy test
ImportError while loading conftest 'C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy\test\conftest.py'.
_pytest.pathlib.ImportPathMismatchError: ('peripy.test.conftest', 'C:\\Users\\zhang\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packag
es\\peripy-1.0.0-py3.9-win-amd64.egg\\peripy\\test\\conftest.py', WindowsPath('C:/Users/zhang/PycharmProjects/PeriPy-master/peripy-master/peripy/
test/conftest.py'))
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master> pytest -cov
============================================================= test session starts ==============================================================
platform win32 -- Python 3.9.5, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master, configfile: ov
collected 0 items / 1 error                                                                                                                     

==================================================================== ERRORS ====================================================================
________________________________________________________ ERROR collecting test session _________________________________________________________
..\..\..\appdata\local\programs\python\python39\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:972: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:972: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:855: in exec_module
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
peripy\__init__.py:6: in <module>
    from .model import Model
peripy\model.py:2: in <module>
    from .integrators import Integrator
peripy\integrators.py:5: in <module>
    from .peridynamics import damage, bond_force, update_displacement, break_bonds
E   ModuleNotFoundError: No module named 'peripy.peridynamics'
=========================================================== short test summary info ============================================================
ERROR  - ModuleNotFoundError: No module named 'peripy.peridynamics'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================================== 1 error in 0.35s ===============================================================
Attached I give my installation records.

1. I used the code platform Pycharm with python 3.9 to install the program in Window 10. Only Installing via wheels (https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyopencl) is successful.
2. •	Install PeriPy pip install peripy
When I executed this code, problem occurs. ValueError: 'XX.pyx' doesn't match any files. The problem is similar to this one.
https://stackoverflow.com/questions/60717795/python-setup-py-sdist-with-cython-extension-pyx-doesnt-match-any-files
I also use “python setup.py install” to avoid this problem.

**Reply by bb515** [reply]
Try running `pytest peripy` from the `peripy` directory, not `peripy_master`, what happens?

**Reply by qianzhang1994** [reply]
### There is no difference.
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master> pytest peripy
ImportError while loading conftest 'C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy\test\conftest.py'.
peripy\__init__.py:6: in <module>
    from .model import Model
peripy\model.py:2: in <module>
    from .integrators import Integrator
peripy\integrators.py:5: in <module>
    from .peridynamics import damage, bond_force, update_displacement, break_bonds
E   ModuleNotFoundError: No module named 'peripy.peridynamics'
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master> cd peripy
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy> pytest peripy
ImportError while loading conftest 'C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy\test\conftest.py'.
__init__.py:6: in <module>
    from .model import Model
model.py:2: in <module>
    from .integrators import Integrator
integrators.py:5: in <module>
    from .peridynamics import damage, bond_force, update_displacement, break_bonds
E   ModuleNotFoundError: No module named 'peripy.peridynamics'
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy>

**Comment by bb515** [comment]
Do you get to choose this device` pyopencl.Device 'Intel(R) Iris(R) Xe Graphics' on 'Intel(R) OpenCL HD Graphics' at 0x19ad52444d0' `
upon running the example2?
and what about `python example1/example.py --opencl`

**Reply by qianzhang1994** [reply]
### There is the full message.
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master> python peripy/examples/example1/example.py
Simulation Progress: 100%|███████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:35<00:00, 27.92steps/s]
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master> python peripy/examples/example1/example.py --opencl
Traceback (most recent call last):
  File "C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy\examples\example1\example.py", line 132, in <module>
    main()
  File "C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master\peripy\examples\example1\example.py", line 86, in main
    integrator = EulerCL(dt=1e-3)
  File "C:\Users\zhang\AppData\Local\Programs\Python\Python39\lib\site-packages\peripy-1.0.0-py3.9-win-amd64.egg\peripy\integrators.py", line 482
, in __init__
    super().__init__(*args, **kwargs)
  File "C:\Users\zhang\AppData\Local\Programs\Python\Python39\lib\site-packages\peripy-1.0.0-py3.9-win-amd64.egg\peripy\integrators.py", line 45,
 in __init__
    raise ContextError
peripy.integrators.ContextError: No suitable context was found. You can manually specifythe context by passing it to ModelCL with the 'context'ar
gument.
PS C:\Users\zhang\PycharmProjects\PeriPy-master\peripy-master>

**Comment by bb515** [comment]
I will suggest changing the code

`peripy/integrators.py::Integrator::__init__::line 42`
`self.context = get_context()`
into
`self.context = cl.create_some_context()`

![image](https://user-images.githubusercontent.com/46718860/138701855-1d1add5f-92e5-4eb7-b60c-349790db0d3d.png)


and then you should be able to choose a device. What devices are listed? What happens when you try each of them?

**Reply by qianzhang1994** [reply]
### Is my device not available?
![image](https://user-images.githubusercontent.com/68472501/138703479-e463def3-e7d0-4a52-bbd4-59d69a1e3332.png)
![image](https://user-images.githubusercontent.com/68472501/138703536-9bae2923-80f3-4e50-b2f4-1799f3345947.png)

**Reply by bb515** [reply]
Thanks for this, what about testing for double_fp_support? Does this raise the ValueError?

![image](https://user-images.githubusercontent.com/46718860/138704305-9b6bc2b6-9be7-44c9-9d8e-d4f7a3257f21.png)

**Reply by qianzhang1994** [reply]
### Yes, it will raise the ValueError
![image](https://user-images.githubusercontent.com/68472501/138705149-0ed50ce1-7de2-46d4-ab10-0ac3806bbb16.png)

**Reply by bb515** [reply]
Okay thanks for this, two things then
1) It seems like your GPU does not support double precision. Single precision has not been tested on PeriPy and so we can't recommend using it as it may lead to numerical instabilities. You could try removing the double precision boiler plate code (at your own risk) to see if it runs a simulations with single precision. If you are looking to use PeriPy for numerical simulations I recommend using a GPU that supports double precision.

2) The correct code behaviour is to raise this `ValueError` for your device when get_context() logic branch is used. I will make an issue to resolve this incorrect code behaviour: not raising the `ValueError`.

**Reply by qianzhang1994** [reply]
Thanks for your help. I will try to find a new device with available GPUs.

**Reply by bb515** [reply]
Thanks for your question that helped us find an issue with the code: https://github.com/alan-turing-institute/PeriPy/issues/118

Closing discussion since it has been resolved.
